worker/containerd/gardenbackend/backend.go

```diff
@@ -3,16 +3,16 @@
 //
 // See https://containerd.io/, and https://github.com/cloudfoundry/garden.
 //
-package backend
+package gardenbackend
```

Comment:
Hm, maybe we could approach it from this angle: what if I'm a new contributor and I need to change something about how Concourse runs containers? [Here's](https://github.com/concourse/concourse/discussions/5495) a specific scenario for that.

I *might* guess `worker/backend`, but like you said, "backend for what?"

I probably *wouldn't* guess `gardenbackend` because I don't even know what the heck Garden is. In fact I would probably guess it's a "garden backend for workers", and not a "backend for garden for workers".

If it's framed purely from a new external contributor standpoint, it may help to reach for higher-level words and/or terminology that we use in documentation.

Reply:
>  what if I'm a new contributor and I need to change something about how Concourse runs containers? 

wdyt of `containers`? sound like something a contributor would go quite quickly towards by seeing `worker` -> `containers` -> `(process_killer.go|network.go|user_namespace.go|etc)`

<details>
<summary>(expand to see an example replacing <code>backend</code>  by <code>containers</code></summary>
<pre>
$ nm $(which concourse) | grep ' t ' | grep backend | sed -e 's/backend/containers/g'
00000000026404b0 t local.github.com/concourse/concourse/worker/containers.(*Backend).BulkInfo
00000000026404e0 t local.github.com/concourse/concourse/worker/containers.(*Backend).BulkMetrics
0000000002640480 t local.github.com/concourse/concourse/worker/containers.(*Backend).Capacity
000000000263fc60 t local.github.com/concourse/concourse/worker/containers.(*Backend).Containers
000000000263eb70 t local.github.com/concourse/concourse/worker/containers.(*Backend).Create
000000000263f520 t local.github.com/concourse/concourse/worker/containers.(*Backend).Destroy
00000000026402e0 t local.github.com/concourse/concourse/worker/containers.(*Backend).GraceTime
0000000002640050 t local.github.com/concourse/concourse/worker/containers.(*Backend).Lookup
000000000263ea80 t local.github.com/concourse/concourse/worker/containers.(*Backend).Ping
000000000263e960 t local.github.com/concourse/concourse/worker/containers.(*Backend).Start
000000000263ea30 t local.github.com/concourse/concourse/worker/containers.(*Backend).Stop
0000000002646510 t local.github.com/concourse/concourse/worker/containers.(*cniNetwork).Add
00000000026414e0 t local.github.com/concourse/concourse/worker/containers.cniNetwork.Add
0000000002646f80 t local.github.com/concourse/concourse/worker/containers.(*CNINetworkConfig).ToJSON
0000000002640510 t local.github.com/concourse/concourse/worker/containers.CNINetworkConfig.ToJSON
00000000026413b0 t local.github.com/concourse/concourse/worker/containers.cniNetwork.generateResolvConfContents
0000000002646660 t local.github.com/concourse/concourse/worker/containers.(*cniNetwork).Remove
00000000026416b0 t local.github.com/concourse/concourse/worker/containers.cniNetwork.Remove
00000000026467b0 t local.github.com/concourse/concourse/worker/containers.(*cniNetwork).SetupMounts
0000000002640d40 t local.github.com/concourse/concourse/worker/containers.cniNetwork.SetupMounts
00000000026427b0 t local.github.com/concourse/concourse/worker/containers.(*Container).Attach
00000000026436b0 t local.github.com/concourse/concourse/worker/containers.(*Container).BulkNetOut
0000000002643430 t local.github.com/concourse/concourse/worker/containers.(*Container).CurrentBandwidthLimits
0000000002643440 t local.github.com/concourse/concourse/worker/containers.(*Container).CurrentCPULimits
0000000002643530 t local.github.com/concourse/concourse/worker/containers.(*Container).CurrentDiskLimits
0000000002643570 t local.github.com/concourse/concourse/worker/containers.(*Container).CurrentMemoryLimits
00000000026437c0 t local.github.com/concourse/concourse/worker/containers.containerdCIO
0000000002641940 t local.github.com/concourse/concourse/worker/containers.(*Container).Handle
00000000026431d0 t local.github.com/concourse/concourse/worker/containers.(*Container).Info
0000000002643230 t local.github.com/concourse/concourse/worker/containers.(*Container).Metrics
0000000002643660 t local.github.com/concourse/concourse/worker/containers.(*Container).NetIn
0000000002643690 t local.github.com/concourse/concourse/worker/containers.(*Container).NetOut
0000000002642e10 t local.github.com/concourse/concourse/worker/containers.(*Container).Properties
0000000002642f40 t local.github.com/concourse/concourse/worker/containers.(*Container).Property
00000000026431b0 t local.github.com/concourse/concourse/worker/containers.(*Container).RemoveProperty
0000000002641b70 t local.github.com/concourse/concourse/worker/containers.(*Container).Run
00000000026432e0 t local.github.com/concourse/concourse/worker/containers.(*Container).SetGraceTime
0000000002643040 t local.github.com/concourse/concourse/worker/containers.(*Container).SetProperty
00000000026419a0 t local.github.com/concourse/concourse/worker/containers.(*Container).Stop
0000000002643290 t local.github.com/concourse/concourse/worker/containers.(*Container).StreamIn
00000000026432b0 t local.github.com/concourse/concourse/worker/containers.(*Container).StreamOut
0000000002647230 t local.github.com/concourse/concourse/worker/containers.(*ErrInvalidInput).Error
0000000002643930 t local.github.com/concourse/concourse/worker/containers.ErrInvalidInput.Error
0000000002647290 t local.github.com/concourse/concourse/worker/containers.(*ErrNotFound).Error
0000000002643950 t local.github.com/concourse/concourse/worker/containers.ErrNotFound.Error
0000000002646910 t local.github.com/concourse/concourse/worker/containers.(*fileStore).Create
00000000026439d0 t local.github.com/concourse/concourse/worker/containers.fileStore.Create
0000000002646a00 t local.github.com/concourse/concourse/worker/containers.(*fileStore).Delete
0000000002643c60 t local.github.com/concourse/concourse/worker/containers.fileStore.Delete
0000000002646440 t local.github.com/concourse/concourse/worker/containers.init
0000000002644280 t local.github.com/concourse/concourse/worker/containers.killer.gracefullyKill
0000000002646b40 t local.github.com/concourse/concourse/worker/containers.(*killer).Kill
0000000002643e50 t local.github.com/concourse/concourse/worker/containers.killer.Kill
0000000002644b30 t local.github.com/concourse/concourse/worker/containers.killer.killProcesses
0000000002644420 t local.github.com/concourse/concourse/worker/containers.killer.killTaskExecedProcesses
0000000002644150 t local.github.com/concourse/concourse/worker/containers.killer.ungracefullyKill
000000000261c3e0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).Containers
000000000261c7f0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).Destroy
000000000261c530 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).GetContainer
000000000261bfc0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).Init
000000000261c1f0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).NewContainer
000000000261c180 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).Stop
000000000261c700 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*client).Version
000000000261d1d0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Checkpoint
000000000261cb00 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Delete
000000000261d0b0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Extensions
000000000261c970 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).ID
000000000261cf10 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Image
000000000261c9d0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Info
000000000261cfa0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Labels
000000000261cba0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).NewTask
000000000261d020 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).SetLabels
000000000261ccf0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Spec
000000000261cdf0 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Task
000000000261d130 t local.github.com/concourse/concourse/worker/containers/libcontainerd.(*container).Update
0000000002645e80 t local.github.com/concourse/concourse/worker/containers.MaxValid
0000000002645cc0 t local.github.com/concourse/concourse/worker/containers.maxValidFromFile
0000000002641880 t local.github.com/concourse/concourse/worker/containers.netNsPath
000000000263e600 t local.github.com/concourse/concourse/worker/containers.New
0000000002640790 t local.github.com/concourse/concourse/worker/containers.NewCNINetwork
0000000002643da0 t local.github.com/concourse/concourse/worker/containers.NewKiller
00000000026459b0 t local.github.com/concourse/concourse/worker/containers.NewRootfsManager
0000000002644ca0 t local.github.com/concourse/concourse/worker/containers.(*Process).ID
0000000002646c20 t local.github.com/concourse/concourse/worker/containers.(*processKiller).Kill
0000000002645100 t local.github.com/concourse/concourse/worker/containers.processKiller.Kill
0000000002644fc0 t local.github.com/concourse/concourse/worker/containers.(*Process).SetTTY
00000000026450e0 t local.github.com/concourse/concourse/worker/containers.(*Process).Signal
0000000002644d00 t local.github.com/concourse/concourse/worker/containers.(*Process).Wait
00000000026436d0 t local.github.com/concourse/concourse/worker/containers.procID
0000000002645740 t local.github.com/concourse/concourse/worker/containers.propertiesToFilterList
0000000002646aa0 t local.github.com/concourse/concourse/worker/containers.(*rootfsManager).SetupCwd
0000000002645a40 t local.github.com/concourse/concourse/worker/containers.rootfsManager.SetupCwd
000000000261d4b0 t local.github.com/concourse/concourse/worker/containers/spec.AllowSyscall
000000000261e750 t local.github.com/concourse/concourse/worker/containers/spec.defaultGardenOciSpec
000000000261d460 t local.github.com/concourse/concourse/worker/containers/spec.deviceWildcard
000000000261e500 t local.github.com/concourse/concourse/worker/containers/spec.envWithDefaultPath
000000000261f2a0 t local.github.com/concourse/concourse/worker/containers/spec.init
000000000261d410 t local.github.com/concourse/concourse/worker/containers/spec.intRef
000000000261ee50 t local.github.com/concourse/concourse/worker/containers/spec.merge
000000000261e2e0 t local.github.com/concourse/concourse/worker/containers/spec.OciResources
000000000261d5e0 t local.github.com/concourse/concourse/worker/containers/spec.OciSpec
000000000261dc20 t local.github.com/concourse/concourse/worker/containers/spec.OciSpecBindMounts
000000000261efc0 t local.github.com/concourse/concourse/worker/containers/spec.rootfsDir
0000000002644610 t local.github.com/concourse/concourse/worker/containers.taskExecedProcesses
0000000002645bd0 t local.github.com/concourse/concourse/worker/containers.(*userNamespace).MaxValidIds
00000000026406e0 t local.github.com/concourse/concourse/worker/containers.WithCNINetworkConfig
00000000026463c0 t local.github.com/concourse/concourse/worker/containers.WithCNINetworkConfig.func1
0000000002640650 t local.github.com/concourse/concourse/worker/containers.WithNameServers
0000000002646360 t local.github.com/concourse/concourse/worker/containers.WithNameServers.func1
000000000263e570 t local.github.com/concourse/concourse/worker/containers.WithNetwork
0000000002646300 t local.github.com/concourse/concourse/worker/containers.WithNetwork.func1
</pre>
</details>

Reply:
Tempting! It would be easy to discover; we publicly use the word "container" in documentation and the CLI. There's also precedent for that sort of name - `atc/builds`, `strings` (Go stdlib), etc.

Reading the type names, they seem to read pretty well - e.g. `containers.WithNetwork` makes sense since it's a container network, not a "backend" network. Maybe we would want to rename the `Backend` type to `GardenBackend` though so it's not confused with "container backend" (which sounds more like what we call a 'runtime').

Reply:
Thanks for the detailed comments!

I like the new contributor angle, and higher level words. Trying to imagine if the new contributor would have read [this](https://concourse-ci.org/internals.html#architecture-worker), and have been using `CONCOURSE_GARDEN` flags by the time they come to look at the worker internals. Probably best to assume they have not as I didn't do that myself :)

Been trying to find examples of packages that use multiple container runtimes for inspiration but the examples may not be relevant to our use case.

cAdvisor uses Docker, crio, and containerd, and separates the code in a straightforward way under a common directory called `container` (like @cirocosta suggested).

Reply:
> Long multi-word package names don't really feel idiomatic.

Agreed. We were trying to figure out a way around this but kept running into the problem of `backend` being too generic. Erred on the side of being explicit there. Also saw other [places](https://github.com/concourse/concourse/blob/3babedbeab761fc601831197f268e04a1e85df22/atc/configvalidate/validate.go#L1) in the code which had double word names but maybe this conversation needed to happen there too. 

Alternatively, we thought of nesting it with a extra directory such as `containerd/garden/backend`, a common pattern in Java but maybe not idiomatic Go.

Reply:
>In the long run, this will be the 'built-in' Concourse runtime, supported along side Kubernetes and whatever other runtimes are developed in the future. Maybe that framing would help come up with a more generic name?

Trying to think through this step by step. 

1. post-Guardian world - Guardian phased out. Still have the `containerd` Garden backend implementation.
2. Kubernetes runtime - We still have this around as the Garden way of doing things needs to phased out.
3. post-Garden world - Garden phased out. The ATC is no longer in charge of scheduling? Or is the ATC using a different interface like [CRI](https://kubernetes.io/blog/2016/12/container-runtime-interface-cri-in-kubernetes/) to talk to the different runtimes? In either case, I imagine the `backend` package would be gone?

Reply:
@xtreme-sameer-vohra no pressure to go through all of this but I believe you had some thoughts on the directory restructuring. feel free to chime in.

Reply:
@muntac

Post-Garden world to me just means we no longer use the Garden client/server/API code, i.e. the boring protocol plumbing and Go API bits. Our `containerd` backend code wouldn't be removed, it would just be adapted to whatever API we come up with to replace Garden and provide the fabled "runtime interface".

It's hard to speak too concretely about the next steps yet as there are still a few open questions in my mind, which perhaps others have thought about already:

* How do we want to support Windows in the future? `containerd` theoretically works on Windows, so do we just do that?
* How do we want to support OS X in the future? Currently we use Garden + Houdini - do we just adapt Houdini to our new API?
* Is this API truly the "runtime interface" used for Linux/Windows/Darwin Concourse workers *and* Kubernetes workers and future runtime implementations? Or do we have a distinction between the two? @cirocosta's demo of a Kubernetes Garden backend seems to indicate that it could span both, at least. 🤔

Reply:
Thanks for those great questions! Still thinking through this. Mostly been trying to concretize many notions that are still abstract in my head. Leaving this here as an update of what I'm working on.

The main question is "how to best to organize the current containerd code for readability and future integrations?" Other questions:

- What kind of functions the 'fabled "runtime interface"' would expose?
- How would the atc use it? 
- How would windows and houdini be supported?
- Kubernetes workers VS using Concourse workers with different runtimes (containerd, rkt, cri-o). Too different or similar enough to conceptualize under a single runtime interface?

To answer those I am or plan to look into following:

- atc/worker client code as I'm only familiar with the standalone worker code to see what it needs from the Garden interface right now.
- @cirocosta's work in #5223 and glancing over https://github.com/concourse/rfcs/pull/20, https://github.com/concourse/rfcs/pull/22, and https://github.com/concourse/concourse/issues/4391
- how Kubelet plugs in different runtimes using [CRI](https://joejulian.name/post/kubernetes-container-engine-comparison/), and if that [interface](https://github.com/kubernetes/cri-api/blob/master/pkg/apis/services.go) along with Garden's can provide any inspiration for Concourse's runtime interface. Can we just use CRI instead of creating our own interface - advantage out of the box support for several runtimes that have native support for that interface.

Going down some fun rabbit holes. Let me know if I'm on the wrong track, and should be diverting energy elsewhere.

Reply:
@vito should we take into account that we might be supporting **multiple runtimes** for whom we have to implement the new interface?
when we move to our own interface in the future would we want to have a **migration period** where we support both Garden and the new interface?
If yes, maybe something like this makes more sense:

worker
|
 —>runtime
 —-—>containerd
 —-———->backend
 —-——————>garden
 —-——————>future-runtime-interface
 ——->other-runtime-A
 —-———->backend
 —-——————>garden
 —-——————>future-runtime-interface

If we don't want to support two interfaces simultaneously:
worker
|
 —>runtime
 —-—>containerd
 —-——--->garden/future-runtime-interface
 —-—>other-runtime-A
 —-——--->garden/future-runtime-interface

Reply:
> when we move to our own interface in the future would we want to have a **migration period** where we support both Garden and the new interface?

In the simplest case, we could swap out all of Garden for something else and just do a major bump of the [worker protocol version](https://github.com/concourse/concourse/blob/3abdcd30813f99244cb1cc1648aa706963e70feb/versions.go#L7-L14).

I think the only time to keep Garden around side-by-side would be during the transition phase which we're already in, where the real intent is to migrate from Guardian to our new `containerd` semantics that we implement in the future (i.e. making privileged containers still be user namespaced). I don't think anyone cares about "Garden support" in particular, so as soon as we can replace Guardian we should be able to replace Garden, too.

All this being said, to be honest this is all just code living in directories and we can move everything around as much as we need to as a clearer picture forms. 🙂 

FWIW, I haven't looked too deeply into CRI but my gut reaction is that it seems too low-level for the type of abstraction we may need in the future - the one that spans "built-in `containerd` / Kubernetes / Nomad / ...".